[Deng Tianle] iP

DengTianle opened the pull request:
## Zazu :bird: The Chat Bot
>“Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Zazu frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,

1. download it.
2. run it.
3. add your tasks.
4. let it manage your tasks for you!

And it is **FREE**!

Features:

 - [X] Managing tasks
 - [X] Managing deadlines
 - [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

DengTianle committed:
Rename, Greet, Exit.

DengTianle committed:
Echo.

DengTianle committed:
Add, List.

DengTianle committed:
Mark as Done

DengTianle committed:
ToDos, Events, Deadlines

DengTianle committed:
Automated Text UI Testing

DengTianle committed:
Handle Errors

DengTianle committed:
Delete

DengTianle committed:
Add saving task list to file using Gson

DengTianle committed:
Stop tracking tasks.json

DengTianle committed:
Update .gitignore to ignore tasks.json

DengTianle committed:
Merge branch 'branch-Level-7'

DengTianle committed:
Add dates and times understanding and bug fixes

DengTianle committed:
Merge branch 'branch-Level-8'

DengTianle committed:
Refactor to use more OOP

DengTianle committed:
Divide classes into packages

DengTianle committed:
Merge add-gradle-support

DengTianle committed:
Automate project builds using Gradle

DengTianle committed:
Add JUnit Tests and bug fixes

DengTianle committed:
Package the App as a JAR file

DengTianle committed:
Add JavaDoc comments

DengTianle committed:
Follow the coding standard

DengTianle committed:
Add find functionality

DengTianle committed:
Merge branch 'branch-A-CodingStandard'

DengTianle committed:
Merge branch 'branch-Level-9'

salads-source commented on src/main/java/zazu/Zazu.java:
```diff
@@ -0,0 +1,123 @@
+public class Zazu {
+
+    /** The task list managed by the application */
+    private TaskList list;
```
Perhaps you could use taskList to explicitly state what kind of tasks it should store?

salads-source commented on src/main/java/zazu/data/task/Deadline.java:
```diff
@@ -0,0 +1,48 @@
+public class Deadline extends Task {
+
+    /** The deadline date by which the task must be completed. */
+    protected LocalDate by;
```
Would changing the deadline name to endTime better to be more explicit?

salads-source commented on src/main/java/zazu/parser/Parser.java:
```diff
@@ -0,0 +1,184 @@
+     * @return A single string formed by joining the elements
+     * of {@code sList} from index {@code start} to {@code end}.
+     */
+    public static String join(String[] sList, int start, int end) {
```
Would changing the method to joinStringArray be better to be more explicit?

salads-source commented on src/main/java/zazu/parser/Parser.java:
```diff
@@ -0,0 +1,184 @@
+     * @return A string representing the command type.
+     * @throws UnknownCommandException If the command is not recognized.
+     */
+    public static String identifyCommand(String str) throws UnknownCommandException {
```
identifyCommand could be slightly ambiguous. Would changing it to extractCommand be better?

salads-source commented on src/main/java/zazu/parser/Parser.java:
```diff
@@ -0,0 +1,184 @@
+     * @return The parsed index (adjusted to zero-based indexing).
+     * @throws IncompleteCommandException If the index is missing from the command.
+     */
+    public static int parseIndex(String str) throws IncompleteCommandException {
```
Would changing it to parseCommandIndex be better for specificity?

salads-source reviewed (changes requested):
LGTM. Codebase mainly adhered to naming conventions with a few small improvements that could be made. Good job!

elizazaa commented on src/main/java/Parser.java:
```diff
@@ -0,0 +1,101 @@
+    public final static String DELETE = "delete";
+    public final static String TODO = "todo";
+    public final static String DEADLINE = "deadline";
+    public final static String EVENT = "event";
```
I like that you made them all final class fields. I think it makes it more readable and easier to debug

elizazaa commented on src/main/java/Parser.java:
```diff
@@ -0,0 +1,101 @@
+        } else if (words[0].equals("deadline")) {
+            return Parser.DEADLINE;
+        } else if (words[0].equals("event")) {
+            return Parser.EVENT;
```
I like the use of parsers here. Will definitely use this for my code as well

elizazaa commented on src/main/java/Zazu.java:
```diff
@@ -1,145 +1,79 @@
 import java.time.format.DateTimeParseException;
```
Would the changes in different classes be better off in separate commits?

elizazaa reviewed (commented):
I like how clean and organized your code is. I learned a lot!

elizazaa commented on src/main/java/Deadline.java:
```diff
@@ -1,14 +1,21 @@
 public class Deadline extends Task {
 
-    protected String by;
+    protected LocalDate by;
```
I like that you use protected instead of the usual private/public modifiers

elizazaa reviewed (commented):
It is easy to read and follow!

DengTianle committed:
Replace CLI with GUI

DengTianle committed:
Merge branch 'branch-Level-10'

DengTianle committed:
Correct build.gradle

Change the java version to the correct version.

DengTianle committed:
Add assertions

Add assertions to ensure that objects are not null unexpectedly. Also,
correct mistake when formatting messages for find command.

DengTianle committed:
Improve code quality

Remove redundant code and magic constants.

DengTianle committed:
Fix style issues

Fix style issues accidentally introduced in the previous commit.

DengTianle committed:
Merge pull request #3 from DengTianle/branch-A-Assertions

Add assertions

DengTianle committed:
Merge branch 'master' into branch-A-CodeQuality

DengTianle committed:
Merge pull request #4 from DengTianle/branch-A-CodeQuality

Branch a code quality

DengTianle committed:
Add task list sorting feature

Add the C-Sort extension which enables sorting the task list in
chronological order.

DengTianle committed:
Merge branch 'branch-BCD-Extension'

DengTianle committed:
Improve the GUI

Display error messages in red color font.

DengTianle committed:
Update readme

Make the readme into a user guide. Upload a screenshot to be used in the
user guide.

DengTianle committed:
Fix readme

Fix problems with alerts in readme.

DengTianle committed:
Fix readme

Fix problems for emojis in readme.

DengTianle committed:
Fix bugs in file storage

Fix bugs that resulted in changes not being saved when user closed the
window rather than using bye command.
